modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/server/ReadOperationForCompactionTracker.java

```diff
@@ -0,0 +1,147 @@
+
+/**
+ * Tracker of read operations from metastorage or its storage. Used to track the completion of read operations before start local
+ * compaction of metastorage.
+ *
+ * <p>Expected usage:</p>
+ * <ul>
+ *     <li>Before starting execution, the reading command invoke {@link #track} with its ID and the compaction revision that is currently
+ *     set ({@link MetaStorageCompactionManager#setCompactionRevisionLocally}/{@link KeyValueStorage#setCompactionRevision}).</li>
+ *     <li>After completion, the reading command will invoke {@link #untrack} with the same arguments as when calling {@link #track},
+ *     regardless of whether the operation was successful or not.</li>
+ *     <li>{@link #collect} will be invoked only after a new compaction revision has been set
+ *     ({@link MetaStorageCompactionManager#setCompactionRevisionLocally}/{@link KeyValueStorage#setCompactionRevision}) for a new
+ *     compaction revision.</li>
+ * </ul>
+ */
+public class ReadOperationForCompactionTracker {
+    private final Map<ReadOperationKey, CompletableFuture<Void>> readOperationFutureByKey = new ConcurrentHashMap<>();
+
+    /**
+     * Starts tracking the completion of a read operation on the current compaction revision.
+     *
+     * <p>Method is expected not to be called more than once for the same arguments.</p>
+     *
+     * <p>Expected usage pattern:</p>
+     * <pre><code>
+     *     Object readOperationId = ...;
+     *     int compactionRevision = ...;
+     *
+     *     tracker.track(readOperationId, compactionRevision);
+     *
+     *     try {
+     *         doReadOperation(...);
+     *     } finally {
+     *         tracker.untrack(readOperationId, compactionRevision);
+     *     }
+     * </code></pre>
+     *
+     * @see #untrack(Object, long)
+     */
+    public void track(Object readOperationId, long compactionRevision) {
+        var key = new ReadOperationKey(readOperationId, compactionRevision);
+
+        CompletableFuture<Void> previous = readOperationFutureByKey.putIfAbsent(key, new CompletableFuture<>());
+
+        assert previous == null : key;
+    }
+
+    /**
+     * Stops tracking the read operation on the compaction revision on which tracking start.
+     *
+     * <p>Method is expected not to be called more than once for the same arguments, and {@link #track} was previously called for same
+     * arguments.</p>
+     *
+     * @see #track(Object, long)
+     */
+    public void untrack(Object readOperationId, long compactionRevision) {
+        var key = new ReadOperationKey(readOperationId, compactionRevision);
+
+        CompletableFuture<Void> removed = readOperationFutureByKey.remove(key);
+
+        assert removed != null : key;
+
+        removed.complete(null);
+    }
+
+    /**
+     * Collects all read operations that were started before {@code compactionRevisionExcluded} and returns a future that will complete
+     * when all collected operations complete.
```

Comment:
```suggestion
     * Returns a future that completes when all read operations that were started before {@code compactionRevisionExcluded} complete.
```

a little inconsistent naming - revisionExcluded and compactionRevisionExcluded in different places